core/kernel/notif.c

```diff
@@ -0,0 +1,224 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, Linaro Limited
+ */
+
+#include <bitstring.h>
+#include <kernel/interrupt.h>
+#include <kernel/mutex.h>
+#include <kernel/notif.h>
+#include <kernel/spinlock.h>
+#include <kernel/thread.h>
+#include <optee_rpc_cmd.h>
+#include <sm/optee_smc.h>
+#include <types_ext.h>
+
+#if CFG_CORE_ASYNC_NOTIF_INTID
+static struct mutex notif_mutex = MUTEX_INITIALIZER;
+static unsigned int notif_lock = SPINLOCK_UNLOCK;
+
+SLIST_HEAD(notif_driver_head, notif_driver);
+static struct notif_driver_head notif_driver_head =
+	SLIST_HEAD_INITIALIZER(&notif_driver_head);
+
+static bitstr_t bit_decl(notif_values, NOTIF_ASYNC_VALUE_MAX + 1);
+static bitstr_t bit_decl(notif_alloc_values, NOTIF_ASYNC_VALUE_MAX + 1);
+static bool notif_started;
+
+TEE_Result notif_alloc_async_value(uint32_t *val)
+{
+	static bool alloc_value_inited;
+	uint32_t old_itr_status = 0;
+	int bit = 0;
+
+	old_itr_status = cpu_spin_lock_xsave(&notif_lock);
+
+	if (!alloc_value_inited) {
+		bit_set(notif_alloc_values, NOTIF_VALUE_DO_BOTTOM_HALF);
+		alloc_value_inited = true;
+	}
+
+	bit_ffc(notif_alloc_values, (int)NOTIF_ASYNC_VALUE_MAX + 1, &bit);
+	if (bit >= 0) {
+		*val = bit;
+		bit_set(notif_alloc_values, bit);
+	}
+
+	cpu_spin_unlock_xrestore(&notif_lock, old_itr_status);
+
+	if (bit < 0)
```

Comment:
maybe a `DMSG()` trace here to help identifying the out-of-mem reporter?

Reply:
I'm not sure that helps much. This is likely to happen when the drivers are initializing so there's bound some more interesting print or perhaps even a panic.